tests/ui/mir/mir_codegen_ssa.rs

```diff
@@ -0,0 +1,19 @@
+// build-pass
+// compile-flags: --crate-type=lib
+#![feature(custom_mir, core_intrinsics)]
+use std::intrinsics::mir::*;
+
+#[custom_mir(dialect = "runtime", phase = "optimized")]
+pub fn f(a: u32) -> u32 {
+    mir!(
+        let x: u32;
+        {
+            // Previously code generation failed with ICE "use of .. before def ..." because the
+            // definition of x was incorrectly identified as dominating the use of x located in the
+            // same statement:
+            x = x + a;
```

Comment:
@tmiasko can you explain what this test tests?

The code here is certainly UB, you can make sure with `rustc +nightly tests/ui/mir/mir_codegen_ssa.rs --emit=llvm-ir --crate-type=lib -O` which generates the following:
```llvm
; mir_codegen_ssa::f
; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable
define noundef i32 @_ZN15mir_codegen_ssa1f17hbe93cec5a8c958a2E(i32 noundef %_1) unnamed_addr #0 {
start:
    ret i32 undef
}
```

But also just notice that this is identical to
```rust
let x;
x = x + a;
return x;
```
Which is an error since `x` is uninitialized on the first use.

Reply:
For some more context I have code locally that produces a spurious unused local variable. It looks like one of the possible fixes is for `successor_within_block` to be removed from the above code. The only failing test is this, and I'm confused on what it tests, as it looks very much invalid...

(for even more context I produce a single terminator which touches a place twice, it assigns to it and then moves out of it and the `successor_within_block` makes the `dominates` call in `visit_local` return `false` (note that the locations are the same, but `successor_within_block` makes the comparison be `a + 1 ≤ a`), creating an additional useless local variable for llvm to chew through)

Reply:
If there is a sensible way to test what you wanted, then such test should replace this; otherwise I'd like to remove `successor_within_block` from above...

Reply:
> The test case has undefined behavior, but it should compile successfully.

Why is that so? I don't think it's possible to produce such MIR without `mir!`?

> A call terminator that moves from a place, and uses the same place as a destination is incorrect

That's not a call terminator though, but rather a new special terminator...

Reply:
> Anyway, feel free to change as necessary.

It actually turns out that I had some wrong assumptions and the right fix is to change my new terminator (it doesn't _actually_ read and write the same place), so no changes are needed here :)